### PR TITLE
VAULT-20669: Add New Authenticated Endpoint for Version 

### PR DESCRIPTION
**Pull request by marcboudreau**
This PR adds a new endpoint `sys/internal/ui/version` which requires a Vault Token that has the **read** capability granted for its path. The change also adds a new rule that grants the **read** capability to this path in the **default** policy. This endpoint is used to obtain the Vault version.

This new path is being introduced to provide a guaranteed mechanism for the Vault UI to retrieve the Vault version, since the current endpoints: `sys/health` and `sys/seal-status` can have the version number redacted if the **redact_version** TCP listener configuration parameter is set to `true`.

The Vault UI will be changed to use this endpoint in a later change.

### CODE REVIEW

**Comment by raskchanky** [comment]
@marcboudreau Neither sys/health nor sys/seal-status are authenticated endpoints. Why does this new endpoint need to be authenticated?

**Comment by raskchanky** [comment]
Is it worth adding a test that verifies this new endpoint outputs the right thing?

**Comment by peteski22** [comment]
> @marcboudreau Neither sys/health nor sys/seal-status are authenticated endpoints. Why does this new endpoint need to be authenticated?

@raskchanky I believe the idea is so that when the Vault platform operator has configured redaction to occur on the unauthenticated endpoints (`sys/health` and `sys/seal-status`) they won't report the version any longer, this PR means there can still be a way to retrieve the version (for the UI) but you have to be logged in to get it.

File: vault/logical_system.go [hunk 1]
```diff
@@ -4754,6 +4754,16 @@ func (b *SystemBackend) pathInternalUIResultantACL(ctx context.Context, req *log
 	return resp, nil
 }
 
+func (b *SystemBackend) pathInternalUIVersion(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
```
Comment (peteski22) [inline vault/logical_system.go:4757]: Please could we add a go doc comment?

File: vault/logical_system_paths.go [hunk 2]
```diff
@@ -2641,6 +2641,31 @@ func (b *SystemBackend) internalPaths() []*framework.Path {
...
+			Pattern: "internal/ui/version",
+			DisplayAttrs: &framework.DisplayAttributes{
+				OperationPrefix: "internal-ui",
+				OperationVerb:   "read",
+				OperationSuffix: "version",
+			},
+			Operations: map[logical.Operation]framework.OperationHandler{
+				logical.ReadOperation: &framework.PathOperation{
+					Callback: b.pathInternalUIVersion,
+					Summary:  "Backwards compatibility is not guaranteed for this API",
```
Comment (peteski22) [inline vault/logical_system_paths.go:2654]: ❓ I wonder why we just use the same summary message for  multiple internal API endpoints, do you think it would be useful to include what this endpoint is for (even if it also says that it doesn't guarantee backwards compatibility)?

**Review by peteski22 — COMMENTED** [review]
Looks good to me, but @raskchanky made a good point in a previous comment that it might be reassuring to have some tests to make sure we get what we're expecting from the endpoint.

Failing that some receipts in the PR summary to show the output? 🤷🏼

**Comment by marcboudreau** [comment]
> Is it worth adding a test that verifies this new endpoint outputs the right thing?

The handler function only has a single branch of execution and so I don't think adding a test for the handler provides much value, however I looked at the version package and there's plenty of untested conditional branching going on there, so I'll add tests there. My thinking is that if we are confident that version.GetVersion().VersionNumber() works correctly, then we can also be confident that the pathInternalUIVersion handler will work as well.

**Comment by raskchanky** [comment]
@peteski22 Thanks for clarifying! Makes sense.

**Comment by marcboudreau** [comment]
My latest two commits add tests for the functions exported from the version package, this will give us the confidence that `version.GetVersion().VersionNumber()` functions correctly. In addition, a line of code inside of an if block that can never be reached has been removed.

**Comment by ccapurso** [comment]
❓ I definitely understand the intent of moving it to an authenticated endpoint. Is there is a purpose to prefixing it with `internal/ui`? The version information used to be generally available in the `sys/seal-status` endpoint. If I understand correctly, we typically reserve `internal/ui` endpoints for those that may fluctuate more heavily than the general system backend APIs.

**Comment by marcboudreau** [comment]
The version information is still available from the unauthenticated endpoints `sys/health` and `sys/seal-status` will continue to report the version, unless the redaction setting is used. This new endpoint is being introduced because the UI needs to retrieve the version (at least once the user is logged in) even if the UI is accessed using a listener (address and port) where the redaction setting is used.

File: version/version.go [hunk 3]
```diff
@@ -23,9 +23,6 @@ func GetVersion() *VersionInfo {
 	if GitDescribe != "" {
 		ver = GitDescribe
 	}
-	if GitDescribe == "" && rel == "" && VersionPrerelease != "" {
-		rel = "dev"
-	}
```
Comment (raskchanky) [inline version/version.go:28]: Why was this removed?

File: version/version_test.go [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+	// Test the git describe case
+	origGitDescribe := GitDescribe
+	GitDescribe = "git-describe"
+	vi = GetVersion()
+	assert.Equal(t, "git-describe", vi.Version)
+
+	GitDescribe = origGitDescribe
+
+	replaceVersion(origVersion, origVersionPrerelease)
```
Comment (raskchanky) [inline version/version_test.go:40]: This pattern feels a little fragile to me. What are your thoughts on having `replaceVersion()` return a closure that resets version information to what it was, and people can run it in a `defer`, similar to how cluster cleanup works?  So something like:

```go
cleanup := replaceVersion("1.2.3", "")
defer cleanup()
```

File: website/content/api-docs/system/internal-ui-version.mdx [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+### Sample response
+
+```json
+{
+  "request_id": "d585b9be-9c6f-a05f-939b-490cf062ebbe",
+  "lease_id": "",
+  "renewable": false,
+  "lease_duration": 0,
+  "data": {
+    "version": "..."
```
Comment (raskchanky) [inline website/content/api-docs/system/internal-ui-version.mdx:42]: I think it would be worth specifying a fake version here, so people know what to expect.

File: version/version.go (hunk 3 above)
Comment (marcboudreau) [reply version/version.go:28]: Because it's unreachable. Right above it, rel is assigned the value of VersionPrerelease and then the condition requires rel to be an empty string and VersionPrerelease to be not empty.

Comment (marcboudreau) [reply version/version.go:28]: I've looked into this and the condition has been like this for 8 years. If no one noticed this in all this time, I really don't think that it's that important.

**Review by raskchanky — APPROVED** [review]

**Review by ccapurso — APPROVED** [review]
Looks good!

**Review by peteski22 — APPROVED** [review]

File: website/content/api-docs/system/internal-ui-version.mdx [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+layout: api
+page_title: /sys/internal/ui/version - HTTP API
+description: >-
+  The `/sys/internal/ui/version` endpoint exposes the software version of Vault.
+---
+
+# `/sys/internal/ui/version`
+
+The `/sys/internal/ui/version` endpoint is used to expose the software version
+of Vault to the UI so that it can be displayed to logged in users.
```
Comment (schavis) [inline website/content/api-docs/system/internal-ui-version.mdx:11]:
```suggestion
The `/sys/internal/ui/version` endpoint exposes the Vault software version
so the Vault UI can display the information to logged in users.
```

Style correction: write in active voice

File: website/content/api-docs/system/internal-ui-version.mdx [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+This is currently only being used internally for the UI. Due to the nature of its
+intended usage, there is no guarantee on backwards compatibility for this endpoint.
```
Comment (schavis) [inline website/content/api-docs/system/internal-ui-version.mdx:14]:
```suggestion
Vault uses internal endpoints to provide information to the Vault UI and
CLI. Internal endpoints are explicitly intended to support Vault
functionality, so we do not recommend them for general use
and do not guarantee backwards compatibility across versions.
```

Style correction: avoid "this" as a pronoun. Make it explicit that we recommend against general use and it's a "use at your own risk" situation

File: website/content/api-docs/system/internal-ui-version.mdx [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+## Get version
+
+This endpoint exposes the software version of Vault to the UI.
```
Comment (schavis) [inline website/content/api-docs/system/internal-ui-version.mdx:18]:
```suggestion
Return the current software version of Vault.
```

Style correction: avoid "this" as a pronoun, technically the endpoint just returns information. The fact that we use it in the UI is secondary

**Review by schavis — CHANGES_REQUESTED** [review]

File: website/content/api-docs/system/internal-ui-version.mdx (hunk 7 above)
Comment (marcboudreau) [reply website/content/api-docs/system/internal-ui-version.mdx:14]: I don't think that there's any practical use by the CLI of the internal endpoints. Are you ok if I take your suggestion minus the `and CLI` (lines 13 and 14).

**Review by schavis — APPROVED** [review]
